src/book/verdant-sundial/software-information-entropy.ipynb

```diff
@@ -0,0 +1,530 @@
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "import math\n",
+    "import pathlib\n",
+    "import shutil\n",
+    "import tempfile\n",
+    "from concurrent.futures import ProcessPoolExecutor, as_completed\n",
+    "\n",
+    "import git\n",
+    "import matplotlib.pyplot as plt\n",
+    "import pandas as pd\n",
+    "import seaborn as sns\n",
+    "\n",
+    "# read example data which includes pubmed github links detected from article abstracts\n",
+    "pd.read_parquet(\"../../../tests/data/examples/pubmed/pubmed_github_links.parquet\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b60573a4",
+   "metadata": {},
+   "source": [
+    "### Lines of Code Changed"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "id": "dc3cb9f6",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def calculate_loc_changes(\n",
+    "    repo_path: pathlib.Path, source: str, target: str, file_names: list[str]\n",
+    ") -> dict[str, int]:\n",
+    "    \"\"\"\n",
+    "    Finds the total number of code lines changed for each specified file between two commits.\n",
+    "\n",
+    "    Args:\n",
+    "        repo_path (pathlib.Path): The path to the git repository.\n",
+    "        source (str): The source commit hash.\n",
+    "        target (str): The target commit hash.\n",
+    "        file_names (list[str]): List of file names to calculate changes for.\n",
+    "\n",
+    "    Returns:\n",
+    "        dict[str, int]: A dictionary where the key is the filename, and the value is the lines changed (added and removed).\n",
+    "    \"\"\"\n",
+    "    repo = git.Repo(repo_path)\n",
+    "    changes = {}\n",
+    "\n",
+    "    for file_name in file_names:\n",
+    "        # Get the diff output for the file between the two commits\n",
+    "        diff_output = repo.git.diff(source, target, \"--numstat\", \"--\", file_name)\n",
+    "        lines_changed = 0\n",
+    "        for line in diff_output.splitlines():\n",
+    "            diff_line = line.split()\n",
+    "            # Check if the line has exactly three parts: (addded,removed,file_path) and that the first two are numeric\n",
+    "            if (\n",
+    "                len(diff_line) == 3\n",
+    "                and diff_line[0].isdigit()\n",
+    "                and diff_line[1].isdigit()\n",
+    "            ):\n",
+    "                added, removed, _ = diff_line\n",
+    "                lines_changed += int(added) + int(removed)\n",
+    "\n",
+    "        changes[file_name] = lines_changed\n",
+    "\n",
+    "    return changes"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2f7c6929",
+   "metadata": {},
+   "source": [
+    "### Normalized Entropy Calculation "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "id": "d715ca03",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def calculate_normalized_entropy(\n",
```

Comment:
Should this be added to the package's modules somewhere if it's not already available (and then imported)? Consider making this a reusable part of the code.